Refactor `Factor` properties

ricardoV94 opened the pull request:
This PR deprecates most of the Factor properties, and gives the following default methods:

* compile_logp(self, vars=None, jacobian=True, sum=True):
* compile_dlogp(self, vars=None, jacobian=True):
* compile_d2logp(self, vars=None, jacobian=True):
 
The prefix `compile` is there to make it obvious this is a costly operation, and that the results should be cached once and reused afterwards. Also this makes it easier to deprecate the old properties, which couldn't easily be changed into methods.

It also deprecates `LoosePointFunc` and renames `FastPointFunc` to `PointFunc` which is now the default. This closes #5318.

Finally this PR also renames `pm.logpt` to `pm.joint_logpt`, which is now the main bridge between `PyMC` and `Aeppl`. The reason for this is that `pm.logpt` has diverged quite a lot from what it was doing in `V3`, and because of that it has a lot of ugly internal logic that should be handled by the `Model` object instead (involving trying to find value variables, transforms and scaling in the tags of the variables). The rename will hopefully help break the dependency that's there for backwards compat and facilitate refactoring in the future, such as #5033.

***
~~It would be nice to find a nice way to transition from `Model.logpt` being a property -> method that can accept `vars`, `jacobian` and `sum` as an argument. Then we can remove the more verbose `logp_elemwiset`, which is temporarily playing this more flexible role, in service of all other methods and properties~~ Decided to break away immediately following discussion in #5333

ricardoV94 committed:
Remove Factor and Loose/FastPoint functions

* compiled functions like `dlogp` are now called `compile_*`
* `logp_elemwiset` -> `logpt(sum=False)`
* `*_no_jacobian` graphs and functions are now obtained via the keyword `jacobian=False`
* Value variables are no longer passed to `model.compile_fn` by default
* Rename `logpt` to `joint_logpt`

ricardoV94 committed:
Convert rvs to values in find_MAP

ricardoV94 committed:
Avoid recompiling [d]logp in every step of find_MAP

This also fixes a bug where graphs with NotImplemented gradients were not detected before deciding on optimization method

michaelosthege commented on pymc/distributions/logprob.py:
```diff
@@ -118,7 +118,7 @@ def _get_scaling(total_size, shape, ndim):
 
 
-def logpt(
+def joint_logpt(
```
Can you do a PR with the same rename on `v3`?
The warning in `3.11.5` points users to the function you're renaming here..

ricardoV94 replied:
For joint_logpt? That one has no good equivalent in V3. Here we are a advising people to use a new V4 model method (logp_elemwise)

joint_logpt should really be only called internally

michaelosthege commented on pymc/model.py:
```diff
@@ -279,57 +280,124 @@ class Factor:
+            "`logp` property has been deprecated, use `compile_logp` instead",
+            FutureWarning,
+        )
         return self.model.fn(self.logpt)
```
Maybe you did this where possible already, but can't more of the properties/methods that now have a warning follow their own instruction and switch to the other API ?

ricardoV94 replied:
I thought that would be more bug prone and difficult to review.

michaelosthege commented on pymc/model.py:
```diff
@@ -1461,7 +1561,7 @@ def __getitem__(self, key):
                 raise e
 
-    def makefn(self, outs, mode=None, *args, **kwargs):
+    def makefn(self, outs, mode=None, **kwargs):
```
```suggestion
    def makefn(self, outs: Sequence[Variable], mode=None, **kwargs) -> Callable([Sequence[np.ndarray]], Sequence[np.ndarray]):
```
Is it something like this?

ricardoV94 replied:
I'll have to check. thanks for the suggestion

michaelosthege commented on pymc/sampling_jax.py:
```diff
@@ -28,7 +28,6 @@
 
 from pymc import Model, modelcontext
 from pymc.backends.arviz import find_observations
-from pymc.distributions import logpt
```
Was this line unused?

ricardoV94 replied:
Let me check again

ricardoV94 replied:
Seems like it was

michaelosthege commented on pymc/tuning/starting.py:
```diff
@@ -113,18 +116,15 @@ def find_MAP(
-                RaveledVars(x, x0.point_map_info)
-            )
-
+        compiled_dlogp_func = DictToArrayBijection.mapf(model.compile_dlogp(rvs, jacobian=False))
+        dlogp_func = lambda x: compiled_dlogp_func(RaveledVars(x, x0.point_map_info))
```
👍

michaelosthege commented on pymc/model.py:
```diff
@@ -788,6 +856,54 @@ def logp_elemwiset(
+        return hessian(cost, value_vars)
+
     @property
     def logpt(self):
```
```suggestion
    def logpt(self) -> Variable:
```
I like the refactor. Adding more type hints will make it easier for beginners like me who can't remember which property is a variable and which one is a function...

michaelosthege reviewed (commented):
Sorry about that nitpick, but given how much code calls these properties/methods, could you add some type hints?
Even incomplete type hints here can give a lot of mypy coverage, right?

ricardoV94 commented on pymc/model.py:
```diff
@@ -788,6 +856,54 @@ def logp_elemwiset(
+                        f"Requested variable {var} not found among the model variables"
+                    )
+
+        cost = self.logpt if not jacobian else self.logp_nojact
```
```suggestion
        cost = self.logpt if jacobian else self.logp_nojact
```

ricardoV94 commented on pymc/model.py:
```diff
@@ -279,57 +280,124 @@ class Factor:
+
+    def compile_d2logp(self, vars=None, jacobian=True):
+        """Compiled log probability density hessian function"""
+        return self.model.compile_fn(self.dlogpt2(vars=vars, jacobian=jacobian))
```
```suggestion
        return self.model.compile_fn(self.d2logpt(vars=vars, jacobian=jacobian))
```

ricardoV94 commented:
Marking as a draft to address #5333 in this PR

twiecki commented on pymc/model.py:
```diff
@@ -279,57 +281,137 @@ class Factor:
+        return self.model.compile_fn(self.d2logpt(vars=vars, jacobian=jacobian))
+
     @property
     def logp(self):
```
So yeah, I would just get rid of all of these.

ricardoV94 committed:
Do not recompute model logpt graph in every step of assign_step_methods

ricardoV94 commented:
Band-aids are now being removed quickly! :adhesive_bandage:

ricardoV94 committed:
Remove unused checkd tests

ricardoV94 committed:
Remove test_model_func.py module

ricardoV94 committed:
Remove old MultiObservedRV tests

Closes #4534

twiecki commented on RELEASE-NOTES.md:
```diff
@@ -39,9 +39,13 @@ All of the above apply to:
 - `pm.sample_prior_predictive`, `pm.sample_posterior_predictive` and `pm.sample_posterior_predictive_w` now return an `InferenceData` object by default, instead of a dictionary (see [#5073](https://github.com/pymc-devs/pymc/pull/5073)).
 - `pm.sample_prior_predictive` no longer returns transformed variable values by default. Pass them by name in `var_names` if you want to obtain these draws (see [4769](https://github.com/pymc-devs/pymc/pull/4769)).
 - `pm.sample(trace=...)` no longer accepts `MultiTrace` or `len(.) > 0` traces ([see 5019#](https://github.com/pymc-devs/pymc/pull/5019)).
+- `logpt`, `logpt_sum`, `logp_elemwiset` and `nojac` variations were removed. Use `Model.logpt(jacobian=jacobian, sum=sum)` instead.
```
```suggestion
- `logpt`, `logpt_sum`, `logp_elemwiset` and `nojac` variations were removed. Use `Model.logpt(jacobian=True/False, sum=True/False)` instead.
```

ricardoV94 replied:
Done

twiecki commented on pymc/model.py:
```diff
@@ -1830,22 +1755,6 @@ def __call__(self, state):
         return self.f(**state)
 
 
-class LoosePointFunc:
```
it's so great that these confusing things finally can go.

twiecki commented:
This is shaping up to be a really nice refactor of some quite obscure code.

ricardoV94 commented:
> This is shaping up to be a really nice refactor of some quite obscure code.

Hopefully one will also be less lost when doing autocomplete on a Model to find a method xD

ricardoV94 committed:
Fix brittle check in test_marginal_likelihood

This would fail when the chains do not have the same length, resulting in the marginal_likelihood being wrapped in a numpy array of `object` dtype. This led the nanmean call to fail with an AttributeError

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -91,6 +91,7 @@ def _filter_parents(self, var, parents) -> Set[VarName]:
 
     def get_parents(self, var: TensorVariable) -> Set[VarName]:
         """Get the named nodes that are direct inputs to the var"""
+        # TODO: Update these lines, variables no longer have a `logpt` attribute
```
Is this broken now?

(Update: This code wasn't working before either.)

ricardoV94 replied:
It was already

michaelosthege commented on pymc/model.py:
```diff
@@ -731,24 +645,53 @@ def logp_dlogp_function(self, grad_vars=None, tempered=False, **kwargs):
+        jacobian: bool = True,
+        sum: bool = True,
+    ):
+        """Compiled log probability density function"""
```
This needs a more detailed docstring.

michaelosthege commented on pymc/model.py:
```diff
@@ -731,24 +645,53 @@ def logp_dlogp_function(self, grad_vars=None, tempered=False, **kwargs):
+        vars: Optional[Union[Variable, Sequence[Variable]]] = None,
+        jacobian: bool = True,
+    ):
+        """Compiled log probability density hessian function"""
```
This one too

michaelosthege commented on pymc/model.py:
```diff
@@ -785,90 +728,101 @@ def logp_elemwiset(
+        self,
+        vars: Optional[Union[Variable, Sequence[Variable]]] = None,
+        jacobian: bool = True,
+    ) -> Variable:
```
Needs a docstring

michaelosthege commented on pymc/model.py:
```diff
@@ -785,90 +728,101 @@ def logp_elemwiset(
+        self,
+        vars: Optional[Union[Variable, Sequence[Variable]]] = None,
+        jacobian: bool = True,
+    ) -> Variable:
```
Also here; the name is not self-explanatory.

michaelosthege commented on pymc/tests/test_shared.py:
```diff
@@ -26,7 +27,10 @@ def test_deterministic(self):
             X = aesara.shared(np.asarray(data_values, dtype=aesara.config.floatX), borrow=True)
             pm.Normal("y", 0, 1, observed=X)
-            model.logp(model.recompute_initial_point())
+            # TODO: This should assert something
```
```suggestion
```

michaelosthege commented on RELEASE-NOTES.md:
```diff
@@ -39,9 +39,13 @@ All of the above apply to:
 - `pm.sample_prior_predictive`, `pm.sample_posterior_predictive` and `pm.sample_posterior_predictive_w` now return an `InferenceData` object by default, instead of a dictionary (see [#5073](https://github.com/pymc-devs/pymc/pull/5073)).
 - `pm.sample_prior_predictive` no longer returns transformed variable values by default. Pass them by name in `var_names` if you want to obtain these draws (see [4769](https://github.com/pymc-devs/pymc/pull/4769)).
 - `pm.sample(trace=...)` no longer accepts `MultiTrace` or `len(.) > 0` traces ([see 5019#](https://github.com/pymc-devs/pymc/pull/5019)).
+- `logpt`, `logpt_sum`, `logp_elemwiset` and `nojac` variations were removed. Use `Model.logpt(jacobian=True/False, sum=True/False)` instead.
```
```suggestion
- `logpt`, `logpt_sum`, `logp_elemwiset` and `nojac` property variations were removed. Use `Model.logpt(jacobian=True/False, sum=True/False)` instead.
```

michaelosthege commented on pymc/model.py:
```diff
@@ -731,24 +645,53 @@ def logp_dlogp_function(self, grad_vars=None, tempered=False, **kwargs):
+        vars: Optional[Union[Variable, Sequence[Variable]]] = None,
+        jacobian: bool = True,
+    ):
+        """Compiled log probability density gradient function"""
```
And here

michaelosthege commented:
I'm open to having docstring link to docstrings of the other properties/methods.
We can probably make the necessary changes all as suggestions and have one commit for them?

ricardoV94 commented:
> I'm open to having docstring link to docstrings of the other properties/methods. We can probably make the necessary changes all as suggestions and have one commit for them?

Yeah I think the `compile_*` is better of just directing to the respective methods for the meaning of the variables. Feel free to make changes and force-push to this branch! Won't have time to work on this until next week

michaelosthege committed:
Add docstrings to various logp related methods
